src/parser.ml

```diff
@@ -1,5 +1,95 @@
 open Ast
 
+module Compat : sig
+  val rep: Uchar.t
+  (** Uchar.rep *)
+
+  val add_utf_8_uchar : Buffer.t -> Uchar.t -> unit
+  (** Buffer.add_utf_8_uchar *)
+
+  val find_opt: ('a -> bool) -> 'a list -> 'a option
+  (** List.find_opt *)
+end = struct
+  let rep: Uchar.t = Obj.magic(0xFFFD)
+  (* let add_utf_8_uchar = Buffer.add_utf_8_uchar *)
+  type buffer =
+   {mutable buffer : bytes;
+    mutable position : int;
+    mutable length : int;
+    initial_buffer : bytes}
+
+  let get_buffer b =
+    let b = (Obj.magic(b): buffer) in
```

Comment:
`Obj.magic` is infamous for being as safe as juggling with chainsaws... so we will have to do something else here. Instead of duplicating part of the `Buffer` internals here, the idea is to simply reimplement `add_utf_8_uchar` below "from the outside" (using `Buffer.add_char`). It should be fairly easy, but shout if you have any questions.